### PR TITLE
C++: Experimental query for implementation of a cryptographic primitive

### PR DESCRIPTION
**Pull request by geoffw0**
Adds an experimental query for detecting custom implementations of a cryptographic primitive (i.e. covering https://mas.owasp.org/MASTG/General/0x04g-Testing-Cryptography/#custom-implementations-of-cryptography
 and parts of https://cwe.mitre.org/data/definitions/1240.html).

Currently `@precision medium`, because it flags established library encryption implementations just as happily as custom implementations and entirely custom algorithms.  Though probably none of these should be in a database unless it's a database for a cryptography library.

Developed during a hackathon.  Needs more work to be production quality - in particular it currently lacks:
- [ ] qhelp / examples --- not strictly necessary for experimental.
- [x] more non-crypto unit tests --- I have added a few more test cases that should not be flagged.
- [x] more crypto unit tests --- I have added a couple of 'homebrew' algorithms in the test that should be flagged.
- [x] and perhaps some more testing in the wild and refinement. --- further refinement: we now reject results that appear to be in encryption libraries.

### CODE REVIEW

File: cpp/ql/src/experimental/Security/CWE/CWE-1240/CustomCryptographicPrimitive.ql [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Implementation of a cryptographic primitive
+ * @description Writing your own cryptographic primitives is prone to errors and omissions that weaken cryptographic protection.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.5
+ * @precision medium
+ * @id cpp/crypto-primitive
+ * @tags security
+ *       external/cwe/cwe-1240
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/experimental/Security/CWE/CWE-1240/CustomCryptographicPrimitive.ql]: Good point.  Added.

**Comment by geoffw0** [comment]
I've just replaced (via a rebase) the examples copied from Wikipedia with code I wrote myself.  This is because I had a chat with some legal + open source people and there was a bit of uncertainty - nobody was 100% sure what I was doing was definitely OK.  In any case the new tests are entirely written by me.  It's slightly sad they're a bit less authentic but they work fine as tests.

Also the latest commit is new - I've added `s?` to the end of a couple of patterns to match names slightly better.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Comment by MathiasVP** [comment]
> (I'm not sure what's wrong with the failing check, but it doesn't appear to be this PR at fault)

Yeah, that's not your fault. It was presumably caused by https://github.com/github/codeql/pull/15003 (which has now been reverted).

**Comment by geoffw0** [comment]
Merging now.  Issue created for taking this query further.

**Pull request merged**